File: src/main/java/com/fengsheng/handler/cheng_qing_save_die_tos.java
package com.fengsheng.handler;

import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.card.Card;
import com.fengsheng.phase.WaitForChengQing;
import com.fengsheng.phase.WaitNextForChengQing;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

public class cheng_qing_save_die_tos extends AbstractProtoHandler<Fengsheng.cheng_qing_save_die_tos> {
    private static final Logger log = Logger.getLogger(cheng_qing_save_die_tos.class);

    @Override
    protected void handle0(HumanPlayer r, Fengsheng.cheng_qing_save_die_tos pb) {
        if (!r.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + r.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        if (!(r.getGame().getFsm() instanceof WaitForChengQing fsm) || r != fsm.askWhom) {
            log.error("现在不是使用澄清的时机");
            return;
        }
        if (!pb.getUse()) {
            r.incrSeq();
            r.getGame().resolve(new WaitNextForChengQing(fsm));
            return;
        }
        Card card = r.findCard(pb.getCardId());
        if (card == null) {
            log.error("没有这张牌");
            return;
        }
        if (card.getType() != Common.card_type.Cheng_Qing) {
            log.error("这张牌不是澄清，而是" + card);
            return;
        }
        Player target = fsm.whoDie;
        if (card.canUse(r.getGame(), r, target, pb.getTargetCardId())) {
            r.incrSeq();
            card.execute(r.getGame(), r, target, pb.getTargetCardId());
        }
    }
}

File: src/main/kotlin/handler/cheng_qing_save_die_tos.kt
package com.fengsheng.handler

import com.fengsheng.HumanPlayer
import com.fengsheng.phase.WaitForChengQing
import com.fengsheng.phase.WaitNextForChengQing
import com.fengsheng.protos.Common
import com.fengsheng.protos.Fengsheng
import com.fengsheng.skill.RuBiZhiShi.excuteRuBiZhiShi
import org.apache.logging.log4j.kotlin.logger

class cheng_qing_save_die_tos : AbstractProtoHandler<Fengsheng.cheng_qing_save_die_tos>() {
    override fun handle0(r: HumanPlayer, pb: Fengsheng.cheng_qing_save_die_tos) {
        if (!r.checkSeq(pb.seq)) {
            logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        if (r.game!!.fsm is excuteRuBiZhiShi) {
            r.game!!.tryContinueResolveProtocol(r, pb)
            return
        }
        val fsm = r.game!!.fsm as? WaitForChengQing
        if (r !== fsm?.askWhom) {
            logger.error("现在不是使用澄清的时机")
            r.sendErrorMessage("现在不是使用澄清的时机")
            return
        }
        if (!pb.use) {
            r.incrSeq()
            r.game!!.resolve(WaitNextForChengQing(fsm))
            return
        }
        val card = r.findCard(pb.cardId)
        if (card == null) {
            logger.error("没有这张牌")
            r.sendErrorMessage("没有这张牌")
            return
        }
        if (card.type != Common.card_type.Cheng_Qing) {
            logger.error("这张牌不是澄清，而是$card")
            r.sendErrorMessage("这张牌不是澄清，而是$card")
            return
        }
        val target = fsm.whoDie
        if (card.canUse(r.game!!, r, target, pb.targetCardId)) {
            r.incrSeq()
            card.execute(r.game!!, r, target, pb.targetCardId)
        }
    }

    companion object {
    }
}